stl/inc/algorithm

```diff
@@ -10204,11 +10204,12 @@ namespace ranges {
                 "The lower bound cannot be greater than the upper bound in a call to std::ranges::clamp "
                 "(N4861 [alg.clamp]/2).");
 
-            if (_STD invoke(_Pred, _STD invoke(_Proj, _Val), _STD invoke(_Proj, _Lo))) {
+            auto&& _Temp = _STD invoke(_Proj, _Val);
+            if (_STD invoke(_Pred, _Temp, _STD invoke(_Proj, _Lo))) {
                 return _Lo;
             }
 
-            if (_STD invoke(_Pred, _STD invoke(_Proj, _Hi), _STD invoke(_Proj, _Val))) {
+            if (_STD invoke(_Pred, _STD invoke(_Proj, _Hi), _STD move(_Temp))) {

```

Comment:
I'm not sure if it is always safe to `move(_Temp)`. When the projection returns a non-const lvalue reference, and the comparer takes its arguments by value, the move constructor could move from the referenced value.

Reply:
Example:
```cpp
string val = "meow";
string low = "m";
string high = "n";
const auto clamped = ranges::clamp(
    ref(val), ref(low), ref(high), [](auto x, auto y) { return x < y; },
    [](const auto& x) -> decltype(auto) { return x.get(); });
assert(val == "meow");
```

Reply:
That is a violation of `regular_invocable`.

Reply:
Sorry for the bad example. As @timsong-cpp pointed out, the code above violate the preconditions of `ranges::clamp`.

Reply:
I think this one was fine actually because the projection returns an lvalue reference. (This is why I deleted my original comment.) The culprit with this one is the unconditional move.

But the case where the projection returns by value (and the comparator takes by value) is a precondition violation.

Reply:
> But the case where the projection returns by value (and the comparator takes by value) is a precondition violation.

This requirement is surprising to me. It would mean that a `regular_invocable` couldn’t move from an xvalue argument, and `func(T)` fails to model `regular_invocable<T>` if `T`’s move constructor isn’t a copy.